### ISSUE TITLE
Train on videos of different length

### ISSUE THREAD

**Issue by kentwhf** [deepmind/kinetics-i3d#97]
Hi,

Thanks for sharing your work first of all. 

I am working on finetuning i3d to my own dataset, which only has two classes. The distribution of the two classes is imbalanced, and I have around 500 video clips in total. One class may account for 80% of the data. I read other posts and open-source materials and found only 64 frames are sampled from each video. In my case, can I train the network with videos of different length in one batch?

I also wonder if it is possible to implement real-time classification with this model? I will be using RGB data only. In my understanding, we have to pre-process data into npy. format, which can be a bit, but not much time-consuming. Eventually, I think the input of RGB images should have similar output as video data. Do you have any advice on this?

**Comment by joaoluiscarreira** [comment]
Hi Kent,

Regarding having variable number of frames in training, this is generally
hard in tensorflow. One possibility is to have a different number of frames
in different batches (e.g. training the first batch with 32, second with
64, etc.) -- tensorflow may be adaptive enough (i did this in tensorflow 2
when experimenting with curriculum training).
Simpler options include cropping the various videos into the same number of
frames as we did for the i3d paper -- that's by far the most used option.
Another option would be to pad the shorter videos to the length of the
longest one.

Regarding real-time classification the simplest approach is to have a delay
of a few frames (e.g. if your model is training on 16 frame clips, then you
could output class labels every 16 frames). If you'd like to run the model
frame by frame, you can either a) pay the computational price by feeding
the previous 16 frames to the model every frame, or b) spend a lot of time
implementing a caching of previous features -- we did this in
https://arxiv.org/abs/1806.03863, it has also been done for wavenet (e.g.
https://arxiv.org/abs/1611.09482).

Best regards,

Joao

**Comment by Kentwhf** [comment]
Thanks for the suggestions!

The length of each video is ranged from 100 to 200-ish frames and the two classes are essentially slipping and not slipping. Considering a slip could happen within a couple of frames, I don't think different batches or cropping suits my case here. I am concern that sub-sampling may exclude desired actions(slips) once original videos are broken to clips (16, or 64 frames etc), then false labels would be assigned to the processed video clips.

I am not sure about the last option. I assumed using the length of the longest video as some parameter of the batch, and then process (pad) shorter videos into one of such length? If so, this may not work well either. 

Another naive question is, can I feed one video to the model one at the time (bach size = 1), so the batch can adapt to videos of different lengths? Or are there other ways that can serve my purpose?

I am still relatively new to this field. Please correct me if I am wrong. :)
